src/dune_rules/action_builder.ml

```diff
@@ -3,6 +3,20 @@ include Dune_engine.Action_builder
 open O
 module With_targets = With_targets
 
+type fail = { fail : 'a. unit -> 'a }
+
+let fail x =
+  let+ () = return () in
+  x.fail ()
+;;
+
+let delayed f =
+  let+ () = return () in
+  f ()
+;;
+
+let of_memo_join f = of_memo f >>= Fun.id
```

Comment:
This new implementation is suboptimal. Perhaps we should introduce `Action_builder.join` to make this a little more natural?

Reply:
Are you worried about an extra `union` with the empty map? It doesn't seem worth optimising to be honest, especially given that we plan to change the implementation soon.

Reply:
Nah, I just find the lack of `val join : 'a t t -> 'a t` to be unnatural for a monadic interface.